File: conformance/run_conformance.py
#!/usr/bin/env python3
"""Ledger Scrolls conformance runner (Python, stdlib only).

Runs the shared fixture corpus in conformance/manifest.json. The same
fixtures are exercised by run_conformance.mjs so the JS and Python
implementations stay byte-compatible.

Usage:
    python3 conformance/run_conformance.py
"""
from __future__ import annotations

import gzip
import hashlib
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent

FAILURES: list[str] = []
PASSES = 0


def check(label: str, ok: bool, detail: str = "") -> None:
    global PASSES
    if ok:
        PASSES += 1
        print(f"  PASS  {label}")
    else:
        FAILURES.append(label)
        print(f"  FAIL  {label}  {detail}")


def sha256_hex(b: bytes) -> str:
    return hashlib.sha256(b).hexdigest()


def canonical_json_bytes(obj) -> bytes:
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")


def read_cbor_len(buf: bytes, pos: int, ai: int):
    if ai < 24:
        return ai, pos
    widths = {24: 1, 25: 2, 26: 4, 27: 8}
    w = widths[ai]
    return int.from_bytes(buf[pos : pos + w], "big"), pos + w


def decode_cbor_bytestring_at(raw: bytes, pos: int = 0) -> tuple[bytes, int]:
    """Decode a CBOR byte string (definite or indefinite length) at pos."""

    mt, ai = raw[pos] >> 5, raw[pos] & 0x1F
    pos += 1
    if mt != 2:
        raise ValueError("not a CBOR byte string")
    if ai == 31:
        out = bytearray()
        while raw[pos] != 0xFF:
            cmt, cai = raw[pos] >> 5, raw[pos] & 0x1F
            if cmt != 2 or cai == 31:
                raise ValueError("invalid chunk")
            clen, pos = read_cbor_len(raw, pos + 1, cai)
            out.extend(raw[pos : pos + clen])
            pos += clen
        return bytes(out), pos + 1
    blen, pos = read_cbor_len(raw, pos, ai)
    return raw[pos : pos + blen], pos + blen


def decode_cbor_bytestring(raw: bytes) -> bytes:
    """Decode standard scroll datum bytes.

    Supports bare CBOR bytes and Plutus constructor-0/tag-121 with one bytes
    field, which is what cardano-cli emits for {"constructor":0,"fields":[...]}.
    """

    if raw[:2] == b"\xd8\x79":
        pos = 2
        if raw[pos] != 0x9F:
            raise ValueError("expected indefinite constructor field array")
        decoded, pos = decode_cbor_bytestring_at(raw, pos + 1)
        if raw[pos] != 0xFF:
            raise ValueError("expected constructor array break")
        return decoded

    return decode_cbor_bytestring_at(raw, 0)[0]


def clean_segment(seg) -> str:
    # Segments appear as plain hex strings or as {"bytes": "<hex>"} objects
    if isinstance(seg, dict):
        seg = seg.get("bytes") or seg.get("seg") or ""
    seg = str(seg).strip()
    return seg[2:] if seg.lower().startswith("0x") else seg


def reconstruct_cip25_pages(metadata_721: dict, policy_id: str) -> bytes:
    policy_meta = metadata_721["721"][policy_id]
    pages = []
    for asset_name, meta in policy_meta.items():
        if meta.get("role") == "manifest" or "MANIFEST" in asset_name.upper():
            continue
        if "payload" not in meta or "i" not in meta:
            continue
        pages.append((int(meta["i"]), meta["payload"]))
    pages.sort(key=lambda p: p[0])
    hex_blob = "".join(clean_segment(s) for _, payload in pages for s in payload)
    raw = bytes.fromhex(hex_blob)
    if raw[:2] == b"\x1f\x8b":
        return gzip.decompress(raw), raw
    return raw, raw


def decode_cbor(raw: bytes, pos: int = 0):
    """Minimal generic CBOR decoder (uint/nint/bytes/text/array/tag) for
    LS-CHAIN manifests. Returns (value, next_pos); tags -> ("tag", n, value)."""

    def read_len(ai, p):
        if ai < 24:
            return ai, p
        w = {24: 1, 25: 2, 26: 4, 27: 8}[ai]
        return int.from_bytes(raw[p : p + w], "big"), p + w

    mt, ai = raw[pos] >> 5, raw[pos] & 0x1F
    if mt == 0:
        return read_len(ai, pos + 1)
    if mt == 1:
        n, p = read_len(ai, pos + 1)
        return -1 - n, p
    if mt in (2, 3):
        n, p = read_len(ai, pos + 1)
        v = raw[p : p + n]
        return (v if mt == 2 else v.decode("utf-8")), p + n
    if mt == 4:
        n, p = read_len(ai, pos + 1)
        out = []
        for _ in range(n):
            v, p = decode_cbor(raw, p)
            out.append(v)
        return out, p
    if mt == 6:
        n, p = read_len(ai, pos + 1)
        v, p = decode_cbor(raw, p)
        return ("tag", n, v), p
    raise ValueError(f"unsupported CBOR major type {mt}")


def reconstruct_chain(manifest_hex: str, pages_by_tx: dict, manifests_by_ptr: dict | None = None):
    """Reconstruct a scroll from its head manifest, following `next` pointers
    (field 7: Constr 0 [] = end, Constr 1 [txHash, ix] = continuation).
    manifests_by_ptr maps "txhash#ix" -> manifest hex for continuation parts."""
    info = None
    hashes = []
    cur = manifest_hex
    parts = 0
    while True:
        m, _ = decode_cbor(bytes.fromhex(cur))
        assert m[0] == "tag" and m[1] == 121, "manifest must be Constr 0"
        f = m[2]
        parts += 1
        if info is None:
            info = {
                "version": f[0], "contentType": f[1].decode(), "codec": f[2].decode(),
                "sizeDecoded": f[3], "sha256Decoded": f[4].hex(), "sha256Encoded": f[5].hex(),
            }
        else:
            assert (f[0] == info["version"] and f[1].decode() == info["contentType"]
                    and f[4].hex() == info["sha256Decoded"]), "continuation file fields mismatch"
        hashes.extend(h.hex() for h in f[6])
        nxt = f[7] if len(f) > 7 else None
        if isinstance(nxt, tuple) and nxt[0] == "tag" and nxt[1] == 122 and nxt[2]:
            key = f"{nxt[2][0].hex()}#{nxt[2][1]}"
            assert manifests_by_ptr and key in manifests_by_ptr, f"continuation manifest missing: {key}"
            cur = manifests_by_ptr[key]
            assert parts <= 32, "manifest chain too long"
        else:
            break
    info["pageTxHashes"] = hashes
    info["parts"] = parts
    encoded = bytearray()
    for tx in info["pageTxHashes"]:
        page = pages_by_tx[tx]["22025"]
        payload = b"".join(bytes.fromhex(clean_segment(s)) for s in page["p"])
        assert sha256_hex(payload) == clean_segment(page["sha"]), "page sha mismatch"
        encoded.extend(payload)
    encoded = bytes(encoded)
    decoded = gzip.decompress(encoded) if info["codec"] == "gzip" else encoded
    return info, encoded, decoded


POINTER_RULES = {
    "utxo-inline-datum-bytes-v1": {
        "txHash": re.compile(r"^[0-9a-fA-F]{64}$"),
        "txIx": int,
    },
    "cip25-pages-v1": {
        "policyId": re.compile(r"^[0-9a-fA-F]{56}$"),
    },
    "url": {
        "url": str,
    },
    # Deprecated aliases readers may still accept
    "utxo-locked-bytes": {
        "txin": re.compile(r"^[0-9a-fA-F]{64}#[0-9]+$"),
    },
    "asset-manifest": {
        "policyId": str,
        "assetName": str,
    },
}


def pointer_is_valid(pointer) -> bool:
    if not isinstance(pointer, dict):
        return False
    rules = POINTER_RULES.get(pointer.get("kind"))
    if rules is None:
        return False
    for field, rule in rules.items():
        value = pointer.get(field)
        if value is None:
            return False
        if isinstance(rule, re.Pattern):
            if not isinstance(value, str) or not rule.match(value):
                return False
        elif not isinstance(value, rule):
            return False
    return True


def main() -> int:
    manifest = json.loads((ROOT / "manifest.json").read_text())
    vectors = manifest["vectors"]

    print("== payload vectors ==")
    for v in vectors["payloads"]:
        data = (ROOT / v["file"]).read_bytes()
        check(f"{v['file']} sha256", sha256_hex(data) == v["sha256"])
        if v.get("codec") == "gzip":
            check(
                f"{v['file']} gunzip sha256",
                sha256_hex(gzip.decompress(data)) == v["decodedSha256"],
            )

    print("== standard scroll datums ==")
    for v in vectors["datums"]:
        raw = bytes.fromhex((ROOT / v["file"]).read_text().strip())
        decoded = decode_cbor_bytestring(raw)
        check(f"{v['file']} decoded sha256", sha256_hex(decoded) == v["decodedSha256"])

    print("== cip25 page reconstruction ==")
    for v in vectors["cip25"]:
        meta = json.loads((ROOT / v["file"]).read_text())
        decoded, gz_raw = reconstruct_cip25_pages(meta, v["policyId"])
        check(f"{v['file']} reconstructed sha256", sha256_hex(decoded) == v["reconstructedSha256"])
        check(f"{v['file']} gzip sha256", sha256_hex(gz_raw) == v["gzipSha256"])

    print("== ls-chain v2 ==")
    for v in vectors.get("chain", []):
        mhex = (ROOT / v["manifest"]).read_text().strip()
        pages = json.loads((ROOT / v["pages"]).read_text())
        conts = {ptr: (ROOT / path).read_text().strip()
                 for ptr, path in (v.get("manifests") or {}).items()}
        info, encoded, decoded = reconstruct_chain(mhex, pages, conts)
        check(f"{v['manifest']} fields", info["contentType"] == v["contentType"]
              and info["codec"] == v["codec"] and len(info["pageTxHashes"]) == v["pageCount"]
              and info["parts"] == v.get("parts", 1))
        check(f"{v['manifest']} encoded sha256", sha256_hex(encoded) == v["encodedSha256"]
              and info["sha256Encoded"] == v["encodedSha256"])
        check(f"{v['manifest']} decoded sha256", sha256_hex(decoded) == v["reconstructedSha256"]
              and info["sha256Decoded"] == v["reconstructedSha256"]
              and len(decoded) == info["sizeDecoded"])

    print("== pointers ==")
    valid_dir = ROOT / vectors["pointers"]["validDir"]
    for f in sorted(valid_dir.glob("*.json")):
        check(f"valid pointer accepted: {f.name}", pointer_is_valid(json.loads(f.read_text())))
    invalid_dir = ROOT / vectors["pointers"]["invalidDir"]
    for f in sorted(invalid_dir.glob("*.json")):
        check(f"invalid pointer rejected: {f.name}", not pointer_is_valid(json.loads(f.read_text())))

    print("== registry canonical hashing ==")
    for v in vectors["registry"]:
        obj = json.loads((ROOT / v["file"]).read_text())
        check(
            f"{v['file']} canonical sha256",
            sha256_hex(canonical_json_bytes(obj)) == v["canonicalSha256"],
        )

    print()
    print(f"{PASSES} passed, {len(FAILURES)} failed")
    return 1 if FAILURES else 0


if __name__ == "__main__":
    sys.exit(main())

File: conformance/run_conformance.mjs
#!/usr/bin/env node
// Ledger Scrolls conformance runner (Node, stdlib only).
// Runs the same fixture corpus as run_conformance.py so the JS and Python
// implementations stay byte-compatible.
//
// Usage: node conformance/run_conformance.mjs

import { createHash } from 'node:crypto';
import { gunzipSync } from 'node:zlib';
import { readFileSync, readdirSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = dirname(fileURLToPath(import.meta.url));

let passes = 0;
const failures = [];

function check(label, ok, detail = '') {
    if (ok) {
        passes++;
        console.log(`  PASS  ${label}`);
    } else {
        failures.push(label);
        console.log(`  FAIL  ${label}  ${detail}`);
    }
}

const sha256Hex = (buf) => createHash('sha256').update(buf).digest('hex');

// Canonical JSON: stable key ordering, no insignificant whitespace.
function canonicalJson(value) {
    if (Array.isArray(value)) return `[${value.map(canonicalJson).join(',')}]`;
    if (value !== null && typeof value === 'object') {
        const keys = Object.keys(value).sort();
        return `{${keys.map(k => `${JSON.stringify(k)}:${canonicalJson(value[k])}`).join(',')}}`;
    }
    return JSON.stringify(value);
}

function readCborLen(raw, pos, ai) {
    if (ai < 24) return [ai, pos];
    const widths = { 24: 1, 25: 2, 26: 4, 27: 8 };
    const w = widths[ai];
    let len = 0n;
    for (let i = 0; i < w; i++) len = (len << 8n) | BigInt(raw[pos + i]);
    return [Number(len), pos + w];
}

// Decode a CBOR byte string (definite or indefinite length) at pos.
function decodeCborBytestringAt(raw, pos = 0) {
    const mt = raw[pos] >> 5, ai = raw[pos] & 0x1f;
    pos++;
    if (mt !== 2) throw new Error('not a CBOR byte string');
    if (ai === 31) {
        const chunks = [];
        while (raw[pos] !== 0xff) {
            const cmt = raw[pos] >> 5, cai = raw[pos] & 0x1f;
            if (cmt !== 2 || cai === 31) throw new Error('invalid chunk');
            const [clen, next] = readCborLen(raw, pos + 1, cai);
            chunks.push(raw.subarray(next, next + clen));
            pos = next + clen;
        }
        return [Buffer.concat(chunks), pos + 1];
    }
    const [blen, next] = readCborLen(raw, pos, ai);
    return [raw.subarray(next, next + blen), next + blen];
}

function decodeStandardDatumBytes(raw) {
    // Bare CBOR bytestring, or Plutus constructor-0/tag-121 with one bytes
    // field from cardano-cli ScriptData JSON.
    if (raw[0] === 0xd8 && raw[1] === 0x79) {
        let pos = 2;
        if (raw[pos] !== 0x9f) throw new Error('expected indefinite constructor field array');
        const [decoded, next] = decodeCborBytestringAt(raw, pos + 1);
        if (raw[next] !== 0xff) throw new Error('expected constructor array break');
        return decoded;
    }
    return decodeCborBytestringAt(raw, 0)[0];
}

const cleanSegment = (seg) => {
    // Segments appear as plain hex strings or as {"bytes": "<hex>"} objects
    if (seg !== null && typeof seg === 'object') seg = seg.bytes ?? seg.seg ?? '';
    seg = String(seg).trim();
    return seg.toLowerCase().startsWith('0x') ? seg.slice(2) : seg;
};

function reconstructCip25Pages(metadata721, policyId) {
    const policyMeta = metadata721['721'][policyId];
    const pages = [];
    for (const [assetName, meta] of Object.entries(policyMeta)) {
        if (meta.role === 'manifest' || assetName.toUpperCase().includes('MANIFEST')) continue;
        if (meta.payload === undefined || meta.i === undefined) continue;
        pages.push([Number(meta.i), meta.payload]);
    }
    pages.sort((a, b) => a[0] - b[0]);
    const hexBlob = pages.flatMap(([, payload]) => payload.map(cleanSegment)).join('');
    const raw = Buffer.from(hexBlob, 'hex');
    if (raw[0] === 0x1f && raw[1] === 0x8b) return [gunzipSync(raw), raw];
    return [raw, raw];
}

// Minimal generic CBOR decoder (uint/nint/bytes/text/array/tag) for
// LS-CHAIN manifests. Returns [value, nextPos]; tags -> {tag, value}.
function decodeCbor(raw, pos = 0) {
    const readLen = (ai, p) => {
        if (ai < 24) return [ai, p];
        const w = { 24: 1, 25: 2, 26: 4, 27: 8 }[ai];
        let n = 0;
        for (let i = 0; i < w; i++) n = n * 256 + raw[p + i];
        return [n, p + w];
    };
    const mt = raw[pos] >> 5, ai = raw[pos] & 0x1f;
    if (mt === 0) return readLen(ai, pos + 1);
    if (mt === 1) { const [n, p] = readLen(ai, pos + 1); return [-1 - n, p]; }
    if (mt === 2 || mt === 3) {
        const [n, p] = readLen(ai, pos + 1);
        const v = raw.subarray(p, p + n);
        return [mt === 2 ? v : Buffer.from(v).toString('utf8'), p + n];
    }
    if (mt === 4) {
        let [n, p] = readLen(ai, pos + 1);
        const out = [];
        for (let i = 0; i < n; i++) { const [v, q] = decodeCbor(raw, p); out.push(v); p = q; }
        return [out, p];
    }
    if (mt === 6) {
        const [n, p] = readLen(ai, pos + 1);
        const [v, q] = decodeCbor(raw, p);
        return [{ tag: n, value: v }, q];
    }
    throw new Error(`unsupported CBOR major type ${mt}`);
}

/* Reconstruct a scroll from its head manifest, following `next` pointers
   (field 7: Constr 0 [] = end, Constr 1 [txHash, ix] = continuation).
   manifestsByPtr maps "txhash#ix" -> manifest hex for continuation parts. */
function reconstructChain(manifestHex, pagesByTx, manifestsByPtr = {}) {
    let info = null;
    const hashes = [];
    let cur = manifestHex;
    let parts = 0;
    for (;;) {
        const [m] = decodeCbor(Buffer.from(cur, 'hex'));
        if (m.tag !== 121) throw new Error('manifest must be Constr 0');
        const f = m.value;
        parts++;
        if (!info) {
            info = {
                version: f[0],
                contentType: Buffer.from(f[1]).toString('utf8'),
                codec: Buffer.from(f[2]).toString('utf8'),
                sizeDecoded: f[3],
                sha256Decoded: Buffer.from(f[4]).toString('hex'),
                sha256Encoded: Buffer.from(f[5]).toString('hex'),
            };
        } else if (f[0] !== info.version
            || Buffer.from(f[1]).toString('utf8') !== info.contentType
            || Buffer.from(f[4]).toString('hex') !== info.sha256Decoded) {
            throw new Error('continuation file fields mismatch');
        }
        for (const h of f[6]) hashes.push(Buffer.from(h).toString('hex'));
        const nxt = f.length > 7 ? f[7] : null;
        if (nxt && nxt.tag === 122 && Array.isArray(nxt.value) && nxt.value.length) {
            const key = `${Buffer.from(nxt.value[0]).toString('hex')}#${nxt.value[1]}`;
            if (!manifestsByPtr[key]) throw new Error('continuation manifest missing: ' + key);
            cur = manifestsByPtr[key];
            if (parts > 32) throw new Error('manifest chain too long');
        } else break;
    }
    info.pageTxHashes = hashes;
    info.parts = parts;
    const chunks = [];
    for (const tx of info.pageTxHashes) {
        const page = pagesByTx[tx]['22025'];
        const payload = Buffer.concat(page.p.map(s => Buffer.from(cleanSegment(s), 'hex')));
        if (sha256Hex(payload) !== cleanSegment(page.sha)) throw new Error('page sha mismatch');
        chunks.push(payload);
    }
    const encoded = Buffer.concat(chunks);
    const decoded = info.codec === 'gzip' ? gunzipSync(encoded) : encoded;
    return [info, encoded, decoded];
}

const POINTER_RULES = {
    'utxo-inline-datum-bytes-v1': { txHash: /^[0-9a-fA-F]{64}$/, txIx: 'number' },
    'cip25-pages-v1': { policyId: /^[0-9a-fA-F]{56}$/ },
    'url': { url: 'string' },
    // Deprecated aliases readers may still accept
    'utxo-locked-bytes': { txin: /^[0-9a-fA-F]{64}#[0-9]+$/ },
    'asset-manifest': { policyId: 'string', assetName: 'string' },
};

function pointerIsValid(pointer) {
    if (pointer === null || typeof pointer !== 'object') return false;
    const rules = POINTER_RULES[pointer.kind];
    if (!rules) return false;
    for (const [field, rule] of Object.entries(rules)) {
        const value = pointer[field];
        if (value === undefined || value === null) return false;
        if (rule instanceof RegExp) {
            if (typeof value !== 'string' || !rule.test(value)) return false;
        } else if (typeof value !== rule) {
            return false;
        }
    }
    return true;
}

const manifest = JSON.parse(readFileSync(join(ROOT, 'manifest.json'), 'utf8'));
const vectors = manifest.vectors;

console.log('== payload vectors ==');
for (const v of vectors.payloads) {
    const data = readFileSync(join(ROOT, v.file));
    check(`${v.file} sha256`, sha256Hex(data) === v.sha256);
    if (v.codec === 'gzip') {
        check(`${v.file} gunzip sha256`, sha256Hex(gunzipSync(data)) === v.decodedSha256);
    }
}

console.log('== standard scroll datums ==');
for (const v of vectors.datums) {
    const raw = Buffer.from(readFileSync(join(ROOT, v.file), 'utf8').trim(), 'hex');
    check(`${v.file} decoded sha256`, sha256Hex(decodeStandardDatumBytes(raw)) === v.decodedSha256);
}

console.log('== cip25 page reconstruction ==');
for (const v of vectors.cip25) {
    const meta = JSON.parse(readFileSync(join(ROOT, v.file), 'utf8'));
    const [decoded, gzRaw] = reconstructCip25Pages(meta, v.policyId);
    check(`${v.file} reconstructed sha256`, sha256Hex(decoded) === v.reconstructedSha256);
    check(`${v.file} gzip sha256`, sha256Hex(gzRaw) === v.gzipSha256);
}

console.log('== ls-chain v2 ==');
for (const v of vectors.chain ?? []) {
    const mhex = readFileSync(join(ROOT, v.manifest), 'utf8').trim();
    const pages = JSON.parse(readFileSync(join(ROOT, v.pages), 'utf8'));
    const conts = Object.fromEntries(Object.entries(v.manifests ?? {})
        .map(([ptr, path]) => [ptr, readFileSync(join(ROOT, path), 'utf8').trim()]));
    const [info, encoded, decoded] = reconstructChain(mhex, pages, conts);
    check(`${v.manifest} fields`, info.contentType === v.contentType
        && info.codec === v.codec && info.pageTxHashes.length === v.pageCount
        && info.parts === (v.parts ?? 1));
    check(`${v.manifest} encoded sha256`, sha256Hex(encoded) === v.encodedSha256
        && info.sha256Encoded === v.encodedSha256);
    check(`${v.manifest} decoded sha256`, sha256Hex(decoded) === v.reconstructedSha256
        && info.sha256Decoded === v.reconstructedSha256
        && decoded.length === info.sizeDecoded);
}

console.log('== pointers ==');
const validDir = join(ROOT, vectors.pointers.validDir);
for (const f of readdirSync(validDir).sort()) {
    const p = JSON.parse(readFileSync(join(validDir, f), 'utf8'));
    check(`valid pointer accepted: ${f}`, pointerIsValid(p));
}
const invalidDir = join(ROOT, vectors.pointers.invalidDir);
for (const f of readdirSync(invalidDir).sort()) {
    const p = JSON.parse(readFileSync(join(invalidDir, f), 'utf8'));
    check(`invalid pointer rejected: ${f}`, !pointerIsValid(p));
}

console.log('== registry canonical hashing ==');
for (const v of vectors.registry) {
    const obj = JSON.parse(readFileSync(join(ROOT, v.file), 'utf8'));
    check(`${v.file} canonical sha256`, sha256Hex(Buffer.from(canonicalJson(obj), 'utf8')) === v.canonicalSha256);
}

console.log();
console.log(`${passes} passed, ${failures.length} failed`);
process.exit(failures.length ? 1 : 0);
